### PR TITLE
Fix: avoid ZSTD codec from overriding service codec factory.

### LINKED ISSUE #7012: [BUG] Zstd new codec is a breaking change for kNN plugin
**Describe the bug**
Plugins cannot continue to use existing approach and register CodecServiceFactory and cluster is crushing in runtime (compile time is ok). Example is kNN plugin, error on cluster start:

https://github.com/opensearch-project/k-NN/actions/runs/4614564115/jobs/8170344936#step:9:1265
```
»  java.lang.IllegalStateException: existing codec service factory already overridden in: org.opensearch.index.codec.customcodecs.CustomCodecPlugin attempting to override again by: org.opensearch.knn.plugin.KNNPlugin
»  	at org.opensearch.index.engine.EngineConfigFactory.<init>(EngineConfigFactory.java:107) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.index.engine.EngineConfigFactory.<init>(EngineConfigFactory.java:63) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.indices.IndicesService.getEngineConfigFactory(IndicesService.java:913) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.indices.IndicesService.createIndexService(IndicesService.java:874) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.indices.IndicesService.createIndex(IndicesService.java:775) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.indices.IndicesService.createIndex(IndicesService.java:209) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.indices.cluster.IndicesClusterStateService.createIndices(IndicesClusterStateService.java:539) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
```

Code on kNN side that is registering CustomServiceFactory:
https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/plugin/KNNPlugin.java#L271-L276

**To Reproduce**
1. Get main branch from k-NN repo
2. Do ```./gradlew run```

**Expected behavior**
There should be alternative way of either enabling/disabling plugins or for plugin to register their CustomServiceFactory.

**Plugins**
distribution build, k-NN is included

**Additional context**
Core change introduced in https://github.com/opensearch-project/OpenSearch/pull/3577 

Part of PR that is registering new factory: https://github.com/opensearch-project/OpenSearch/pull/3577/files#diff-629964cdbeae1039851179519fc3e3d3cc45e8329d208dd9f2b7034537675262R38

Core code where the check for only one factory is happening: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java#L79

### PR DESCRIPTION
**Pull request by mulugetam**
The new [ZSTD compression codec](https://github.com/opensearch-project/OpenSearch/pull/3577/) adds ZSTD to the existing compression codecs (`default`, `best_compression`, and `lucene_default`). This PR allows the compression codec to give a custom-codec service factory only when `index.codec` is set to `ZSTD` or `ZSTDNODICT`.

### Description
Fixes issue #7012 by explicitly avoiding the creation of a custom codec service unless the `index.codec` value is either `ZSTD` or `ZSTDNODICT` .

### Issues Resolved
Resolves #7012 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dblock — COMMENTED** [review]
This works, but feels awfully specific to the fact that we have these two custom codecs in the project. A few ideas before I hit approve/merge.

1. Can we check whether the codec is _not_ custom (aka well known codecs) instead of whether it's a known custom codec?
2. Does it make sense to make names such as "CUSTOM:ZSTD" and then look for "CUSTOM:" instead or is it a silly idea?

File: sandbox/plugins/custom-codecs/src/main/java/org/opensearch/index/codec/customcodecs/CustomCodecPlugin.java [hunk 1]
```diff
@@ -26,15 +27,25 @@
...
 
     /**
      * @param indexSettings is the default indexSettings
-     * @return the engine factory
+     * @return the custom-codec service factory
      */
     @Override
     public Optional<CodecServiceFactory> getCustomCodecServiceFactory(final IndexSettings indexSettings) {
-        return Optional.of(new CustomCodecServiceFactory());
+        String codec = indexSettings.getValue(EngineConfig.INDEX_CODEC_SETTING);
```
Comment (navneet1v) [inline sandbox/plugins/custom-codecs/src/main/java/org/opensearch/index/codec/customcodecs/CustomCodecPlugin.java:45]: please add unit test for this class.

**Comment by mulugetam** [comment]
> 1. dec is _not_ custom

@dblock Do you mean requiring the user to specify all custom codecs as, for example, `index.codec: CUSTOM:ZSTD` instead of just `index.codec: ZSTD`? And then return a custom CodecServiceFactory only `if(codec.startsWith('CUSTOM:'))`?

**Comment by navneet1v** [comment]
One comment, if an index sets this value to use the custom codec and from another plugin the codec is coming lets say a k-NN index, which codec will be picked up? or it will lead to failures?

Example(might result in failure, can we check this case):

```
PUT my-knn-index-1
{
  "settings": {
    "index": {
      "knn": true,
      "knn.algo_param.ef_search": 100,
      "codec": "ZSTD"
    }
  },
  "mappings": {
    "properties": {
        "my_vector1": {
          "type": "knn_vector",
          "dimension": 2,
          "method": {
            "name": "hnsw",
            "space_type": "l2",
            "engine": "nmslib",
            "parameters": {
              "ef_construction": 128,
              "m": 24
            }
          }
        },
        "my_vector2": {
          "type": "knn_vector",
          "dimension": 4,
          "method": {
            "name": "hnsw",
            "space_type": "innerproduct",
            "engine": "faiss",
            "parameters": {
              "ef_construction": 256,
              "m": 48
            }
          }
        }
    }
  }
}
```

I have created this issue: https://github.com/opensearch-project/OpenSearch/issues/7032 tries to explore the possible solution. 

One suggestion I have is can we add java doc on top of this plugins and also on the EnginePlugin class which provides this interface that if an index tries to use 2 codec this can lead to failures in creating the index.

**Comment by mulugetam** [comment]
I reviewed the code again to see if I could do it without overriding the existing code service factory, and I can. So, I have removed the CodecService classes altogether. 

Note that the custom compression codecs are registered by calling [org.apache.lucene.codecs.Codec](https://github.com/apache/lucene/blob/c31017589bf60d94b587087545b582be2064dacf/lucene/core/src/java/org/apache/lucene/codecs/Codec.java#L73) ctor [here](https://github.com/opensearch-project/OpenSearch/blob/a76ed954468fac9a6443ad3e19856cf6e84fd7d4/sandbox/plugins/custom-codecs/src/main/java/org/opensearch/index/codec/customcodecs/Lucene95CustomCodec.java#L95). The `index.codec` setting will get the custom compression codecs by calling `Codec.availableCodecs().contains(s)` [here](https://github.com/opensearch-project/OpenSearch/blob/fe8b4d4f1db61659407be6537427e0729da79794/server/src/main/java/org/opensearch/index/engine/EngineConfig.java#L132) in EngineConfig.

@dblock The names for custom compression codecs, `ZSTD` and `ZSTDNODICT`, cannot contain non-alphanumeric characters; Lucene's [NamedSPILoader](https://github.com/apache/lucene/blob/c31017589bf60d94b587087545b582be2064dacf/lucene/core/src/java/org/apache/lucene/util/NamedSPILoader.java#L82) forbids it.

@navneet1v @martin-gaievski Can you check if this also addresses #7032?

**Comment by navneet1v** [comment]
> I reviewed the code again to see if I could do it without overriding the existing code service factory, and I can. So, I have removed the CodecService classes altogether.

Now you have removed all CodecSeviceFactory , I want to know how the codec is now getting used or getting attached to a particular index?

is it like if someone specify the index.codec: ZSTD the ZstdCodec would be picked up if already present?

**Comment by mulugetam** [comment]
> is it like if someone specify the index.codec: ZSTD the ZstdCodec would be picked up if already present?

@navneet1v Yes. 

Lucene's `setCodec(new ZstdCodec())`  is called, just like `setCodec(new Lucene95Codec(Lucene95Codec.Mode.BEST_COMPRESSION))`  is called for a `index.codec: best_compression`.

**Comment by reta** [comment]
> I reviewed the code again to see if I could do it without overriding the existing code service factory, and I can. So, I have removed the CodecService classes altogether.

This is awesome, @mulugetam , basically the standard service loader mechanism is purely sufficient here, rigth?

**Review by reta — APPROVED** [review]

**Comment by mulugetam** [comment]
> This is awesome, @mulugetam , basically the standard service loader mechanism is purely sufficient here, right?
@reta yes.

**Comment by mulugetam** [comment]
@reta I have also upgraded the zstd-jni version from 1.5.4-1 to 1.5.5-1. Version 1.5.5-1 is based on ZSTD version 1.5.5 that addresses the rare corruption bug described here: https://github.com/facebook/zstd/releases/tag/v1.5.5

**Review by dblock — APPROVED** [review]

**Pull request merged**